core/src/connection_reuse.rs

```diff
@@ -40,193 +40,543 @@
+    }
+}
+
+/// Struct shared between most of the `ConnectionReuse` infrastructure.
+/// Knows about all connections and their current state, allows one to poll for
+/// incoming and outbound connections, while managing all state-transitions that
+/// might occur automatically.
+// #[derive(Clone)]
+struct ConnectionsManager<T, C>
+where
+    T: Transport,
     C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
-    C::Output: StreamMuxer,
+    C::Output: StreamMuxer + Clone,
 {
-    // Underlying transport and connection upgrade for when we need to dial or listen.
-    inner: UpgradedNode<T, C>,
+    /// Underlying transport and connection upgrade, used when we need to dial or listen.
+    transport: UpgradedNode<T, C>,
+
+    /// All the connections that were opened, whether successful and/or active or not.
+    // TODO: this will grow forever
+    connections: Mutex<FnvHashMap<Multiaddr, PeerState<C::Output>>>,
+
+    /// Tasks to notify when one or more new elements were added to `connections`.
+    notify_on_new_connec: Mutex<HashMap<usize, task::Task, UsizeHasherBuilder>>,
 
-    // Struct shared between most of the `ConnectionReuse` infrastructure.
-    shared: Arc<Mutex<Shared<C::Output>>>,
+    /// Counter giving us the next listener_id
+    listener_counter: Counter,
 }
 
-struct Shared<M>
+impl<T, C> ConnectionsManager<T, C>
 where
-    M: StreamMuxer,
+    T: Transport,
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
+    C::Output: StreamMuxer + Clone,
 {
-    // List of active muxers.
-    active_connections: FnvHashMap<Multiaddr, M>,
+    /// Notify all `Tasks` in `self.notify_on_new_connec` that a new connection was established
+    /// consume the tasks in that process
+    fn new_con_notify(&self) {
+        let mut notifiers = self.notify_on_new_connec.lock();
+        trace!("Notify {} listeners about a new connection", notifiers.len());
+        for (_, task) in notifiers.drain() {
+            task.notify();
+        }
+    }
 
-    // List of pending inbound substreams from dialed nodes.
-    // Only add to this list elements received through `add_to_next_rx`.
-    next_incoming: Vec<(M, M::InboundSubstream, Multiaddr)>,
+    /// Register `task` under id if not yet registered
+    fn register_notifier(&self, id: usize, task: task::Task) {
+        let mut notifiers = self.notify_on_new_connec.lock();
+        notifiers.insert(id, task);
+    }
 
-    // New elements are not directly added to `next_incoming`. Instead they are sent to this
-    // channel. This is done so that we can wake up tasks whenever a new element is added.
-    add_to_next_rx: mpsc::UnboundedReceiver<(M, M::InboundSubstream, Multiaddr)>,
+    /// Clear the cached connection if the entry contains an error. Returns whether an error was
+    /// found and has been removed.
+    fn clear_error(&self, addr: Multiaddr) -> bool {
+        let mut conns = self.connections.lock();
+
+        if let Entry::Occupied(e) = conns.entry(addr) {
```

Comment:
Should be `get()` instead of `entry()` so that `clear_error` can take `addr` by reference.